IPython/core/display.py

```diff
@@ -140,6 +141,88 @@ def _new_id():
     return b2a_hex(os.urandom(16)).decode('ascii')
 
 
+class _RecursiveObject:
+    """
+    Default recursive object that provides a recursion repr if needed.
+    """
+
+    def __repr__(self):
+        return '<recursion ...>'
+
+    def _repr_html_(self):
+        return '&lt;recursion ...&gt;'
+
+
+DataMetadata = namedtuple('DataMetadata', ['data', 'metadata'])
+
+_objs = set()
+
+
+
+def get_repr(obj, include=None, exclude=None, *, on_recursion=_RecursiveObject):
+    """
+    return the representations of an object and associated metadata.
+
+    An given object can have many representation available, that can be defined
+    in many ways: `_repr_*_` methods, `_repr_mimebundle_`, or user-registered
+    formatter for types.
+
+    When given an object, `get_repr()` will search for the various formatting
+    option with their associated priority and return the requested
+    representation and associated metadata.
+
+
+    Parameters
+    ----------
+    obj : an objects
+        The Python objects to get the representation data.
+    include : list, tuple or set, optional
+        A list of format type strings (MIME types) to include in the
+        format data dict. If this is set *only* the format types included
+        in this list will be computed.
+    exclude : list, tuple or set, optional
+        A list of format type strings (MIME types) to exclude in the format
+        data dict. If this is set all format types will be computed,
+        except for those included in this argument.
+    on_recursion: callable
+        Return an object to compute the representation when recursion is
+        detected.
+
+
+    Returns
+    -------
+    (format_dict, metadata_dict) : named tuple of two dicts
+
+        0/data: format_dict is a dictionary of key/value pairs, one of each
+        format that was generated for the object. The keys are the format
+        types, which will usually be MIME type strings and the values and
+        JSON'able data structure containing the raw data for the
+        representation in that format.
+
+        1/metadata: metadata_dict is a dictionary of metadata about each
+        mime-type output. Its keys will be a strict subset of the keys in
+        format_dict.
+
+    Note
+    ----
+
+    When `get_repr` detect it is recursively called, it will attempt to return
+    the representation of `_RecursiveObject`
+
+    """
+    global _objs
```

Comment:
I don't see any way that does not require a global-ish state, I can move it into a class (that would be better).
But we can't really enforce users to pass variable down. 

I'm still expecting this to almost never go on and recurse. 
If you are concern about threading I can use  `threading.local()`. Coroutines: there is no tasklocals in Python (yet?) and having a task-switch would require having coroutiness based reprs, which I'm unsure we do support. 

Worse case is I believe we will go down the recursion path where there is no recursions, (AFAICAT), and I think it is a reasonable tradeoff.